core/modules/replica/HttpIngestModule.h

```diff
@@ -0,0 +1,300 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_HTTPINGESTMODULE_H
+#define LSST_QSERV_HTTPINGESTMODULE_H
+
+// System headers
+#include <memory>
+#include <string>
+
+// Third party headers
+#include "nlohmann/json.hpp"
+
+// Qserv headers
+#include "replica/HttpModule.h"
+#include "util/Mutex.h"
+
+// Forward declarations
+namespace lsst {
+namespace qserv {
+namespace replica {
+    class DatabaseInfo;
+namespace database {
+namespace mysql {
+    class Connection;
+}}}}} // Forward declarations
+
+// This header declarations
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+/**
+ * Class HttpIngestModule provides a support ingesting catalogs
+ * into Qserv.
+ */
+class HttpIngestModule: public HttpModule {
+public:
+
+    typedef std::shared_ptr<HttpIngestModule> Ptr;
+
+    static Ptr create(Controller::Ptr const& controller,
+                      std::string const& taskName,
+                      unsigned int workerResponseTimeoutSec);
+
+    HttpIngestModule() = delete;
+    HttpIngestModule(HttpIngestModule const&) = delete;
+    HttpIngestModule& operator=(HttpIngestModule const&) = delete;
+
+    ~HttpIngestModule() final = default;
+
+protected:
+
+    /**
+     * Supported values for parameter 'subModuleName':
+     *
+     *   TRANSACTIONS              for many transactions (possible selected by various criteria)
+     *   SELECT-TRANSACTION-BY-ID  for a single transaction
+     *   BEGIN-TRANSACTION         for starting a new transaction
+     *   END-TRANSACTION           for finishing/aborting a transaction
+     *   ADD-DATABASE              for adding a new database for the data ingest
+     *   PUBLISH-DATABASE          for publishing a database when data ingest is over
+     *   DELETE-DATABASE           for deleting an unpublished (partially ingest) database
+     *   ADD-TABLE                 for adding a new table for the data ingest
+     *   ADD-CHUNK                 for registering (or requesting a status of) of a new chunk
+     *   BUILD-CHUNK-LIST          for building (or rebuilding) an "empty chunk list"
```

Comment:
FWIW I'm not _totally_ buying this string-based submodule dispatch thing. 

I think an enum might be better?

Reply:
I understand. When I was designing this "mini-framework" I went through a similar process by weighting "pros" and "const" of `enum` vs `name`. A problem with `enum` is that I would have to make then sub-class-specific. For example, if I had some module:
```
HttpMyModule : public HttpModule {
public:
    typedef td::shared<HttpMyModule> Ptr;
    enum SubModule {
        GetSomeSomethingDone,
        PullInfo,
        ..
    };
protected:
    void execureImpl(SubModule subModule, ... ) final;
    ..
};
```
Then I would have to store a pointer to that module in a header of class `HttpProcessor` by its final class, something along these lines:
```
class HttProcessor {
...
private:
    HttpMyModule::Ptr const _myModule;
};
```
Right now I'm not exposing final classes of modules in that header:
```
    HttpModule::Ptr const _myModule;
};
```
Exposing the final classes through the header would violate one of the fundamental principle of the OOP which is known as the "open-closed" principle:
https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle